### PR TITLE
[Pages][ALM] Site-referenced table discovery + dependency-aware solution splitting

### PR DESCRIPTION
**Pull request by tyaginidhi**
**PR 3 of 4 (stacked):** plan-only → EDM → **table-discovery** → refresh-enforcement. Targets the PR-B branch (`users/nityagi/edm-site-support`); review/merge after #191 and #192.

## Problem

Two coupled defects in `plan-alm`'s size/split analysis:

1. **Table discovery over-counted via publisher prefix.** Prefix-matching `EntityDefinitions` catches every table sharing the prefix. With a shared/default publisher (`new_`, env default) a **6-table site matched 22 unrelated tables** — and it simultaneously *missed* the site's real tables when they came from a different prefix (e.g. a `bp_*` template under an `edm` publisher).
2. **The split heuristic produced ~one solution per table.** The old schema-segmentation grouped by table-name stem, so a correctly-counted 20-table site still fanned out into ~20 solutions (the absurd 21-solution split that surfaced this).

## Fix

**Site-referenced table discovery** (`resolve-site-tables.js`): scope custom tables to what the site's `.powerpages-site/table-permissions/` (+ datamodel manifest) actually reference, intersected with the env's custom-unmanaged tables. SME-confirmed signal: *"If a table is used in the site there will be permissions for it."* `tableCountScope` ∈ `site-referenced | manifest-only | unavailable` (never an env-wide prefix dump). `--publisherPrefix` now scopes **only** the env var count.

**Dependency-aware, capacity-bounded splitting** (`compute-split-plan.js`): build connected-component clusters from table relationships (lookups + N:N) via union-find, then first-fit-decreasing bin-pack **whole clusters** (never splitting a relationship) into the fewest solutions under `maxTableCount`/`maxSchemaAttrs`, capped at `maxSchemaSplitSolutions` (default 8). An indivisible over-cap cluster stays whole + raises an oversized-cluster recommendation. **The split trigger + thresholds are unchanged — only the packing.**

## New shared libs

- `resolve-site-tables.js` — site-referenced table scoping (single source of truth)
- `query-metadata.js` — consolidated custom-unmanaged-table query (was triplicated)
- `query-table-relationships.js` — relationship edges as a lib; `audit-permissions` CLI is now a thin wrapper over it (+ ManyToMany)
- `validation-helpers.js` — `odataGet` / `odataGetAll` shared `@odata.nextLink` paginator

`estimate-solution-size.js` emits `tableCountScope` + `tableRelationships[]`; `compute-split-plan.js` consumes the edges; `setup-solution` Phase 5.2.D uses the shared discovery helper instead of a prefix dump.

## Verification

- `node --test plugins/power-pages/scripts/tests/` → **1209 pass / 0 fail**
- `node plugins/power-pages/scripts/lint-skills-alm.js` → **0 findings**
- Live (EDM site): discovery returns the `bp_*` set, not the env-wide prefix list; 6 tables → single solution, 34 tables → 3 solutions.
- Plugin version bumped 2.3.0 → 2.4.0.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by tyaginidhi** [comment]
@copilot resolve the merge conflicts in this pull request

File: plugins/power-pages/scripts/lib/compute-split-plan.js [hunk 1]
```diff
@@ -319,9 +319,12 @@ function partitionBySchema(estimate, meta, config) {
...
     sizeMB: round(sizePerDomain),
-    componentCount: Math.ceil(
-      (estimate.schemaAttrCount || 0) / domainCount,
-    ),
+    // A Table domain's component count IS its table count when known (each table
+    // is one Entity solution component). Falls back to an even attr-share split
+    // only for explicit domains that didn't list their tables.
+    componentCount: (dom.tableLogicalNames && dom.tableLogicalNames.length > 0)
+      ? dom.tableLogicalNames.length
+      : Math.ceil((estimate.schemaAttrCount || 0) / domainCount),
```
[the remark this answers is not in the document]
Comment (tyaginidhi) [reply plugins/power-pages/scripts/lib/compute-split-plan.js]: Fixed in bb6f7f7. Table-domain `componentCount` is now a schema-component proxy — `sum(attributeCount) + 1 per table` (entity component) when the domain's tables are known — instead of the raw table count. This stops an over-cap Table solution slipping past `validateSplits`' maxComponentCount check and stops distorting the Site solution's subtracted count. Test asserts 500+300+2 = 802.

File: plugins/power-pages/scripts/lib/compute-split-plan.js [hunk 2]
```diff
@@ -704,6 +785,17 @@ function computeSplitPlan({ estimate, config, meta }) {
...
+  // per-solution cap means a single connected dependency cluster couldn't be
+  // split without cutting a relationship. Name it so the user can decide whether
+  // to denormalize the schema or raise the cap — we never silently split a cluster.
+  const oversizedClusterWarnings = proposedSolutions
+    .filter((s) => Array.isArray(s.tableLogicalNames) &&
+      s.tableLogicalNames.length > config.thresholds.maxTableCount)
+    .map((s) => ({
+      type: 'warning',
+      message: `Solution ${s.uniqueName} holds ${s.tableLogicalNames.length} related tables — above the ${config.thresholds.maxTableCount}-per-solution cap — because they form one dependency cluster that cannot be split without breaking a relationship. Consider denormalizing the schema or raising maxTableCount in .alm-config.json.`,
+    }));
```
[the remark this answers is not in the document]
Comment (tyaginidhi) [reply plugins/power-pages/scripts/lib/compute-split-plan.js:814]: Fixed in bb6f7f7. Added a companion oversized-SCHEMA guard: any Table solution whose summed `attributeCount` exceeds `maxSchemaAttrs` now raises a recommendation. This catches the ceiling-boundary case (>`maxSchemaSplitSolutions` independent attr-heavy clusters → FFD least-loaded fallback co-locates them). Regression test added with 9×14000-attr independent clusters; AGENTS.md packing claim corrected.

File: plugins/power-pages/scripts/lib/validation-helpers.js [hunk 3]
```diff
@@ -236,6 +236,53 @@ function makeRequest({ url, method = 'GET', headers = {}, body = null, includeHe
...
+async function odataGetAll(url, token, request = makeRequest, maxPages = 100) {
+  const out = [];
+  let next = url;
+  for (let p = 0; p < maxPages && next; p++) {
+    const page = await odataGet(next, token, request);
+    if (Array.isArray(page.value)) out.push(...page.value);
+    next = page['@odata.nextLink'] || null;
+  }
+  return out;
+}
```
[the remark this answers is not in the document]
Comment (tyaginidhi) [reply plugins/power-pages/scripts/lib/validation-helpers.js:295]: Fixed in bb6f7f7. `odataGetAll` now FAILS CLOSED — it throws when it exits at the `maxPages` cap with `@odata.nextLink` still present, rather than silently returning a truncated set (which would skew ALM table/env-var counts with no signal). Callers wanting partial results must catch + downgrade intentionally. Test added.

File: plugins/power-pages/scripts/lib/resolve-site-tables.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      let records = [];
+      try { records = loadTablePermissions(dir); } catch { records = []; }
+      for (const r of records) {
+        const name = r && r.entitylogicalname;          // NOT entityname (that's the display label)
+        if (typeof name === 'string' && name.trim()) {
+          names.add(name.trim().toLowerCase());
+          sources.tablePermissions += 1;
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (tyaginidhi) [reply plugins/power-pages/scripts/lib/resolve-site-tables.js:68]: Fixed in bb6f7f7. `sources.tablePermissions` now counts the `*.tablepermission.yml` FILES present (up-front), not parsed records — so a temporarily-malformed file no longer drops the count to 0 and misclassifies `tableCountScope` as manifest-only. Test added with a malformed permission file (sources.tablePermissions=1, names empty).

**Comment by tyaginidhi** [comment]
@copilot resolve the merge conflicts in this pull request

File: plugins/power-pages/scripts/lib/compute-split-plan.js [hunk 5]
```diff
@@ -312,16 +312,32 @@ function partitionBySchema(estimate, meta, config) {
...
+  // counting 1-per-table severely undercounts and lets an over-cap Table solution
+  // slip past validateSplits' maxComponentCount check (and distorts the Site
+  // solution's count, which subtracts the domain counts). Proxy = sum(attributeCount)
+  // + 1 per table (the entity component). attributeCount comes from estimate.tables[].
+  const attrByTable = new Map(
+    (Array.isArray(estimate.tables) ? estimate.tables : [])
+      .map((t) => [t && t.logicalName, (t && t.attributeCount) || 0]),
+  );
+  const schemaComponentProxy = (names) =>
+    (names || []).reduce((sum, n) => sum + (attrByTable.get(n) || 0), 0) + (names ? names.length : 0);
```
[the remark this answers is not in the document]
Comment (tyaginidhi) [reply plugins/power-pages/scripts/lib/compute-split-plan.js]: Fixed. `attrByTable` (the schema-component proxy) now lowercases both its keys and the lookup names, so it matches case-insensitively even when user-authored .alm-config.json domains differ from estimate.tables' Dataverse casing. Regression test added (bp_Inspection vs bp_inspection → 802, not 2).

File: plugins/power-pages/scripts/lib/compute-split-plan.js [hunk 6]
```diff
@@ -704,6 +798,37 @@ function computeSplitPlan({ estimate, config, meta }) {
...
+  // attr-heavy table clusters must share the capped number of split solutions, the
+  // FFD packer's least-loaded fallback co-locates clusters and a bucket busts the
+  // column cap. Without this, the overflow is silent (the table-count guard alone
+  // misses it). attributeCount comes from estimate.tables[].
+  const attrByTable = new Map(
+    (Array.isArray(estimate.tables) ? estimate.tables : [])
+      .map((t) => [t && t.logicalName, (t && t.attributeCount) || 0]),
+  );
+  const solutionSchemaAttrs = (s) =>
+    (s.tableLogicalNames || []).reduce((sum, n) => sum + (attrByTable.get(n) || 0), 0);
```
[the remark this answers is not in the document]
Comment (tyaginidhi) [reply plugins/power-pages/scripts/lib/compute-split-plan.js]: Fixed in the same change — the oversized-schema (attr-cap) guard's `attrByTable`/`solutionSchemaAttrs` are now lowercased too, so a casing mismatch can't undercount and silently miss the overflow warning.

File: plugins/power-pages/skills/setup-solution/SKILL.md [hunk 7]
```diff
@@ -304,14 +304,18 @@ GET {envUrl}/api/data/v9.2/powerpagesitelanguages?$filter=_powerpagesiteid_value
...
 
-1. Read `.datamodel-manifest.json` if present (for the known list of tables created by `setup-datamodel`)
-2. **Also** query the environment directly for all custom unmanaged tables, filtering by the publisher prefix:
-```
-GET {envUrl}/api/data/v9.2/EntityDefinitions?$select=LogicalName,MetadataId,IsManaged,IsCustomEntity
+> **Why not publisher prefix:** prefix-matching over-counts catastrophically with a shared/default publisher (`new_`, env default) — a 6-table site can match 22 unrelated tables — and it also *misses* the site's real tables when they come from a different prefix (e.g. a `bp_*` template under an `edm` publisher). The authoritative signal (SME-confirmed) is the site's **table permissions**: "If a table is used in the site there will be permissions for it."
+
+Run the shared discovery helper, which scopes custom tables to the site's table permissions (+ datamodel manifest) intersected with the env's custom-unmanaged tables:
+```bash
+node "${CLAUDE_PLUGIN_ROOT}/scripts/lib/discover-site-components.js" \
```
[the remark this answers is not in the document]
Comment (tyaginidhi) [reply plugins/power-pages/skills/setup-solution/SKILL.md]: Fixed. Changed `${CLAUDE_PLUGIN_ROOT}` → `${PLUGIN_ROOT}` at the Phase 5.2.D discover-site-components call. The repo uses `${PLUGIN_ROOT}` post Open Plugins migration (56 files; this file uses it everywhere else) — the .mcp.json/hooks.json CLAUDE_PLUGIN_ROOT refs are deliberate Claude Code runtime config and stay.

**Review by priyanshu92 — APPROVED** [review]
